README.md

```diff
@@ -40,9 +40,10 @@ The `pkg-fun.nix` files in all unit directories are automatically discovered, ca
 
 These requirements will be checked using CI:
 1. The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${substring 0 4 name}/${name}`.
-2. <a id="req-ref"/> Files outside a unit directory must not reference files inside that unit directory, and the other way around.
+2. <a id="req-ref-out"/> Files inside a unit directory must not reference files outside that unit directory.
+3. <a id="req-ref-in"/> Files outside a unit directory must not reference files inside a unit directory, except for definitions of variant attributes¹ in `all-packages.nix` and the auto-calling logic.
 4. The definition of a package in the unit directory is the one `pkgs.<name>` points to.
-5. To avoid problems with merges, if a package attribute is defined by a unit directory, it must not be defined in `pkgs/top-level/all-packages.nix` or `pkgs/top-level/aliases.nix`.
+5. To avoid problems with merges, if a package attribute is defined by a unit directory, an attribute of the same name in `pkgs/top-level/all-packages.nix` (or `pkgs/top-level/aliases.nix`) must not redefine it in terms of something other than the unit.
```

Comment:
```suggestion
```

This appears to be a duplicate of 4 now?

Reply:
Yeah looks like it, might have even been a duplicate before that. I also think we can avoid mentioning `aliases.nix`, that can be part of a more general check that overlays don't redefine the same attributes. Here's my suggestion to replace 3, 4 and 5:

3. Files outside a unit directory must not reference files inside a unit directory except for the auto-calling logic and the next rule
4. If `pkgs/top-level/all-packages.nix` defines the same attribute that is also defined by the unit directory `<unitDir>`, then its definition must be of the form `pkgs.callPackage <unitDir> <args>`

Reply:
Actually, what should the restriction be here? Because it doesn't make sense to just allow `pkgs.callPackage <unitDir> <args>`, because then we might as well just allow `all-packages.nix` to redefine attributes from unit directories, then you can just use `hello = super.hello.override { ... }`. (Yes there's the rare use case of having to re-call the file to avoid infinite recursion, I don't think this would warrant such a special case though)

On the other hand, if we allow e.g. `callPackages`, `pythonPackages.callPackage`, `libsForQt5.callPackage`, `import` etc. to be used, it would lead to `pkg-fun.nix` not being as standard anymore (it might not return a package, it might get arguments that don't correspond to top-level attributes, it might not have `.override` anymore), which is also not great. It _would_ allow using unit directories for pretty much all definitions though, but we decided against that before.

Reply:
> Actually, what should the restriction be here?

None. We already have four reasonable ones.

In the meeting we did discuss a check along the lines of
1. gather the list of all attribute names that have a corresponding unit `pkg-fun.nix`
2. call `all-packages.nix` such that `callPackage` returns a special value containing the path to `pkg-fun.nix`
3. intersect (1) and (2) and check that they match

Maybe that's what you're trying to think of?


> Because it doesn't make sense to just allow `pkgs.callPackage <unitDir> <args>`

This is not a problem.


> you can just use `hello = super.hello.override { ... }`.

No, because this is evaluating too much. The automatic `{ }` args may produce an exception or some other result that isn't an attribute set that `callPackage` can `//` an `override` attribute onto.
Concretely, this means that for example performance suffers and packages that need bootstrapping are excluded.

Reply:
In [meeting #26](https://discourse.nixos.org/t/2022-01-23-nixpkgs-architecture-team-meeting-26/24892) we agreed that restricting it to just `pkgs.callPackage`, and that the `.override` indeed doesn't work all the time. @roberth could attempt an implementation of that.

Let's merge this for now and I'll suggest further changes with a separate PR